Source/JavaScriptCore/heap/FreeListInlines.h

```diff
@@ -33,35 +33,45 @@ namespace JSC {
 template<typename Func>
 ALWAYS_INLINE HeapCell* FreeList::allocate(const Func& slowPath)
 {
-    unsigned remaining = m_remaining;
-    if (remaining) {
-        unsigned cellSize = m_cellSize;
-        remaining -= cellSize;
-        m_remaining = remaining;
-        return bitwise_cast<HeapCell*>(m_payloadEnd - remaining - cellSize);
+    unsigned cellSize = m_cellSize;
+    if (LIKELY(m_payloadStart != m_payloadEnd)) {
+        char* result = m_payloadStart;
+        m_payloadStart += cellSize;
+        return bitwise_cast<HeapCell*>(result);
     }
     
-    FreeCell* result = head();
-    if (UNLIKELY(!result))
+    FreeCell* cell = head();
+    if (isSentinel(cell))
         return slowPath();
+
+    auto [offsetToNext, lengthInBytes] = cell->decode(m_secret);
+    m_head = bitwise_cast<FreeCell*>(bitwise_cast<char*>(m_head) + offsetToNext);
+    m_payloadStart = bitwise_cast<char*>(cell);
+    m_payloadEnd = m_payloadStart + lengthInBytes;
     
-    m_scrambledHead = result->scrambledNext;
+    // We assume we don't create any empty intervals; so there should be enough space remaining
+    // to allocate a cell.
+    char* result = m_payloadStart;
+    m_payloadStart += cellSize;
     return bitwise_cast<HeapCell*>(result);
 }
 
 template<typename Func>
 void FreeList::forEach(const Func& func) const
 {
-    if (m_remaining) {
-        for (unsigned remaining = m_remaining; remaining; remaining -= m_cellSize)
-            func(bitwise_cast<HeapCell*>(m_payloadEnd - remaining));
-    } else {
-        for (FreeCell* cell = head(); cell;) {
-            // We can use this to overwrite free objects before destroying the free list. So, we need
-            // to get next before proceeding further.
-            FreeCell* next = cell->next(m_secret);
-            func(bitwise_cast<HeapCell*>(cell));
-            cell = next;
+    FreeCell* cell = head();
+    char* payloadStart = m_payloadStart;
+    char* payloadEnd = m_payloadEnd;
+    ASSERT(payloadEnd - payloadStart < (ptrdiff_t)(16 * KB));
+    while (payloadStart < payloadEnd) {
+        for (; payloadStart < payloadEnd; payloadStart += m_cellSize)
+            func(bitwise_cast<HeapCell*>(payloadStart));
+        if (!isSentinel(cell)) {
+            auto [offsetToNext, lengthInBytes] = cell->decode(m_secret);
+            payloadStart = bitwise_cast<char*>(cell);
+            payloadEnd = payloadStart + lengthInBytes;
+            ASSERT(payloadEnd - payloadStart < (ptrdiff_t)(16 * KB));
+            cell = bitwise_cast<FreeCell*>(bitwise_cast<char*>(cell) + offsetToNext);
```

Comment:
Just use `payloadStart` instead of `bitwise_cast<char*>(cell)` here.

Reply:
Resolved in latest push.